content/blog/2019-09-25-type_checker.md

```diff
@@ -0,0 +1,100 @@
+
+   Note: We haven't taken jumps / branches into consideration which implies if there was a jump instruction before where the variable was defined, we would still throw an error. In hindsight, this can be resolved by traversing the basic blocks in a topological order.
+
+
+
+## Hardest parts during the implementation
+
+The type checker implementation, though straightforward, had a few challenges.
+1. The typing rules need to be designed carefully for each operation, including arithmetic, boolean, and control instructions. Take the following branch instruction as an example: br b left right. The type checker needs to go through all the existing boolean variables to ensure that b is predefined and left and right are valid labels in the code snippet. Also, though currently only two basic types are supported in bril, there could be extensions for more types like list, stack etc. Thus, it is important to maintain modularity of arithmetic/boolean and control flow checking so that future updates could be made easily to support them.
+2. To ensure that the type checker works properly with labels, a first pass of the labels is designed. Through the first pass, all existing labels are saved and no duplicate ones are allowed to avoid conflicts. Though this almost doubles the type checking overhead, it further ensures the correctness of the program by static analysis.
+3. Keeping track of the line number and returning proper error message when encountered with an type checking error. To track the line number, the type checker works on the Bril text representation and is developed with its own parser. When scanning through the code line by line after the first pass, a line number is maintained during type checking to help the programmer debug. In order to maintain the line number with blank lines and comments, we take in the original code snippet and ignore these lines while updating it during type checking.
+
+
+
+## Evaluation and results
+
+We wrote a set of benchmarking programs to test our implementation for various types of rules defined in the design section above.
+The general test cases are classified into two sub-directories: [should-fail](https://github.com/tissue3/bril/tree/master/test/type-check/should-fail) and [should-pass](https://github.com/tissue3/bril/tree/master/test/type-check/should-pass). It has expected output named as `*.out` corresponding to the input file `*.bril`. User can simply run all test cases in a directory by running `turnt directory/*.bril`. For example, to run should-fail benchmark, one can just run `turnt test/type-check/should-pass/*.bril` at the main directory of bril.
+Because should-pass cases are trivial, where we enumerate list existing operations, in the following table we only list tests cases that our type checker will report an error message. The second column of the table aims to help one understand why an error would be reported by our type checker.
```

Comment:
Looks like you missed this comment about "enumerate list"? https://github.com/sampsyo/cs6120/pull/34#discussion_r331818904